File: typescript/runner.ts
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// Copyright (c) 2026, Victor Duarte

import { CallbackContext, InvalidOutcome } from './context.js'

import type { Intent } from './context.js'
import type {
  Action,
  Completed,
  EndTerminal,
  ExitTerminal,
  Failed,
  Failure,
  FailureKind,
  RunHandle,
  RunResult,
  ScopeFailure,
  ScopeResult,
  Terminal,
} from './contracts.js'
import type { CompiledPlacement, CompiledScope, CompiledSnapshot } from './graph.js'

const missing = Symbol('missing')

export function startRuntime<State extends object>(snapshot: CompiledSnapshot, state: State): RunHandle<State> {
  return new Handle(() => new Run(snapshot, state).execute())
}

class Run<State extends object> {
  readonly snapshot: CompiledSnapshot
  readonly state: State
  nextActivationId = 1
  nextFailureId = 0
  nextScopeId = 0
  nextTerminalSequence = 0

  constructor(snapshot: CompiledSnapshot, state: State) {
    this.snapshot = snapshot
    this.state = state
  }

  async execute(): Promise<RunResult<State>> {
    const outcome = await this.runScope(1, 1, undefined)
    if (outcome.failed) {
      return Object.freeze<Failed<State>>({
        status: 'failed',
        state: this.state,
        terminals: outcome.terminals,
        failure: outcome.failure,
      })
    }
    return Object.freeze<Completed<State>>({ status: 'completed', state: this.state, terminals: outcome.terminals })
  }

  private async runScope(compiledScopeId: number, ownerActivationId: number, input: unknown): Promise<ScopeSuccess | ScopeFailed> {
    const scope = this.scope(compiledScopeId)
    const runtimeScopeId = ++this.nextScopeId
    let queue = [this.activation(scope.entryElementId, input)]
    const terminals: Terminal[] = []
    let started = 0
    let failure: FailureSignal | undefined

    while (queue.length > 0 && failure === undefined) {
      if (scope.maxActivations !== undefined && started >= scope.maxActivations) {
        const blocked = queue[0]!
        failure = this.signal(
          this.failure(
            'activation_limit',
            `Flow ${JSON.stringify(scope.name)} exceeded maxActivations`,
            runtimeScopeId,
            blocked.activationId,
            blocked.elementId,
          ),
          blocked.activationId,
          blocked.input,
        )
        break
      }

      let width = Math.min(scope.concurrency, queue.length)
      if (scope.maxActivations !== undefined) width = Math.min(width, scope.maxActivations - started)
      const batch = queue.slice(0, width)
      queue = queue.slice(width)
      started += batch.length
      const outcomes = await Promise.all(batch.map((activation) => this.runActivation(scope, runtimeScopeId, activation)))

      for (const outcome of outcomes) {
        if (outcome.failed) {
          terminals.push(...outcome.terminals)
          failure ??= outcome
          continue
        }
        for (const item of outcome.items) {
          if (item.type === 'next') {
            if (failure === undefined) queue.push(this.activation(item.elementId, item.input))
          } else {
            terminals.push(item)
          }
        }
      }
    }

    const ordered = Object.freeze(terminals.slice().sort((left, right) => left.sequence - right.sequence))
    if (failure !== undefined) {
      return this.recoverScope(scope, runtimeScopeId, ownerActivationId, failure, ordered)
    }
    if (scope.combine === undefined) return { failed: false, terminals: ordered }

    const result = this.scopeResult(ordered)
    const callback = await this.callback(scope.combine, input, result)
    if (!callback.ok) {
      const signal = this.callbackFailure(
        callback.cause,
        'flow_combine',
        'Flow combine failed',
        runtimeScopeId,
        ownerActivationId,
        scope.ownerElementId,
        null,
        input,
        result,
      )
      return this.recoverScope(scope, runtimeScopeId, ownerActivationId, signal, ordered)
    }
    if (callback.intents.length === 0) return { failed: false, terminals: ordered }

    const transformed = this.boundaryTerminals(callback.intents, ownerActivationId)
    const invalid = this.invalidRootExit(scope, transformed, runtimeScopeId)
    if (invalid !== undefined) {
      return this.recoverScope(scope, runtimeScopeId, ownerActivationId, invalid, ordered)
    }
    return { failed: false, terminals: transformed }
  }

  private async runActivation(scope: CompiledScope, runtimeScopeId: number, activation: Activation): Promise<Success | FailureSignal> {
    const placement = this.placement(activation.elementId)
    if (placement.kind === 'flow') {
      if (placement.ownedScopeId === undefined) throw new Error('compiled Flow has no owned scope')
      // ponytail: recurse by Flow depth; use an explicit stack only if real
      // workflows reach the JavaScript call-stack limit.
      const child = await this.runScope(placement.ownedScopeId, activation.activationId, activation.input)
      if (child.failed) return this.signal(child.failure, activation.activationId, activation.input, null, child.terminals)
      return this.routeChild(scope, runtimeScopeId, placement, activation, child.terminals)
    }

    const intents = await this.runNode(placement, runtimeScopeId, activation)
    if (intents.failed) return intents
    return this.routeIntents(scope, runtimeScopeId, placement, activation, intents.intents)
  }

  private async runNode(
    placement: CompiledPlacement,
    scopeId: number,
    activation: Activation,
  ): Promise<IntentSuccess | FailureSignal> {
    if (placement.handler === undefined || placement.retry === undefined) {
      throw new Error('compiled Node is incomplete')
    }
    let previous: Failure | null = null

    for (let attempt = 1; attempt <= placement.retry.maxAttempts; attempt++) {
      const callback = await this.callback(placement.handler, activation.input)
      if (callback.ok) {
        const intents =
          callback.intents.length === 0
            ? [{ kind: 'emit', action: null, value: activation.input, hasValue: true } satisfies Intent]
            : callback.intents
        return { failed: false, intents }
      }

      const kind: FailureKind = callback.cause instanceof InvalidOutcome ? 'invalid_outcome' : 'handler'
      const failure = this.failure(
        kind,
        message(callback.cause),
        scopeId,
        activation.activationId,
        placement.elementId,
        attempt,
        callback.cause,
        previous,
      )
      if (kind === 'handler' && attempt < placement.retry.maxAttempts) {
        const policy = this.retry(placement, attempt, failure)
        if (typeof policy !== 'number') {
          if (policy !== null) return this.signal(policy, activation.activationId, activation.input)
        } else {
          if (policy > 0) await new Promise((resolve) => setTimeout(resolve, policy))
          previous = failure
          continue
        }
      }
      return this.recoverNode(placement, scopeId, activation, failure)
    }
    throw new Error('retry loop did not return')
  }

  private retry(placement: CompiledPlacement, attempt: number, failure: Failure): number | null | Failure {
    if (placement.retry === undefined) throw new Error('compiled Node has no retry policy')
    try {
      const shouldRetry = placement.retry.shouldRetry(failure)
      if (typeof shouldRetry !== 'boolean') throw new TypeError('shouldRetry must return boolean')
      if (!shouldRetry) return null
      const configured = placement.retry.delayMs
      const delay = typeof configured === 'function' ? configured(attempt, failure) : configured
      if (!Number.isSafeInteger(delay) || delay < 0) throw new TypeError('delayMs must return a nonnegative safe integer')
      return delay
    } catch (cause) {
      return this.failure(
        'retry_policy',
        message(cause),
        failure.scopeId,
        failure.activationId,
        failure.elementId,
        failure.attempt,
        cause,
        failure,
      )
    }
  }

  private async recoverNode(
    placement: CompiledPlacement,
    scopeId: number,
    activation: Activation,
    failure: Failure,
  ): Promise<IntentSuccess | FailureSignal> {
    if (placement.recover === undefined) return this.signal(failure, activation.activationId, activation.input)
    const callback = await this.callback(placement.recover, activation.input, failure)
    if (!callback.ok) {
      const kind: FailureKind = callback.cause instanceof InvalidOutcome ? 'invalid_outcome' : 'node_recovery'
      return this.signal(
        this.failure(
          kind,
          message(callback.cause),
          scopeId,
          activation.activationId,
          placement.elementId,
          null,
          callback.cause,
          failure,
        ),
        activation.activationId,
        activation.input,
      )
    }
    if (callback.intents.length === 0) return this.signal(failure, activation.activationId, activation.input)
    return { failed: false, intents: callback.intents }
  }

  private async recoverScope(
    scope: CompiledScope,
    runtimeScopeId: number,
    ownerActivationId: number,
    signal: FailureSignal,
    terminals: readonly Terminal[],
  ): Promise<ScopeSuccess | ScopeFailed> {
    if (scope.recover === undefined) return { failed: true, terminals, failure: signal.failure }
    const scopeFailure = Object.freeze<ScopeFailure>({
      primary: signal.failure,
      terminals,
      result: signal.result,
      failingActivationId: signal.activationId,
    })
    const callback = await this.callback(scope.recover, signal.input, scopeFailure)
    if (!callback.ok) {
      const kind: FailureKind = callback.cause instanceof InvalidOutcome ? 'invalid_outcome' : 'flow_recovery'
      return {
        failed: true,
        terminals,
        failure: this.failure(
          kind,
          message(callback.cause),
          runtimeScopeId,
          ownerActivationId,
          scope.ownerElementId,
          null,
          callback.cause,
          signal.failure,
        ),
      }
    }
    if (callback.intents.length === 0) return { failed: true, terminals, failure: signal.failure }

    const transformed = this.boundaryTerminals(callback.intents, ownerActivationId)
    const invalid = this.invalidRootExit(scope, transformed, runtimeScopeId)
    if (invalid !== undefined) return { failed: true, terminals, failure: invalid.failure }
    return { failed: false, terminals: transformed }
  }

  private async callback(callback: Function, input: unknown, extra: unknown | typeof missing = missing): Promise<CallbackResult> {
    const context = new CallbackContext(this.state, input)
    try {
      const value = await (extra === missing ? callback(context) : callback(context, extra))
      if (value !== undefined) throw new InvalidOutcome('callbacks must return undefined')
      return { ok: true, intents: context.close() }
    } catch (cause) {
      context.close()
      return { ok: false, cause }
    }
  }

  private routeIntents(
    scope: CompiledScope,
    scopeId: number,
    placement: CompiledPlacement,
    activation: Activation,
    intents: readonly Intent[],
  ): Success | FailureSignal {
    const destinations: { readonly intent: Intent; readonly target: number | undefined }[] = []
    for (const intent of intents) {
      if (intent.kind === 'end') {
        destinations.push({ intent, target: undefined })
        continue
      }
      const target = placement.links.find((link) => link.action === intent.action)?.targetElementId
      if (target === undefined && intent.action !== null && !scope.exits.includes(intent.action)) {
        return this.signal(
          this.failure(
            'unknown_action',
            `unknown action ${JSON.stringify(intent.action)} from ${JSON.stringify(placement.name)}`,
            scopeId,
            activation.activationId,
            placement.elementId,
          ),
          activation.activationId,
          activation.input,
        )
      }
      destinations.push({ intent, target })
    }

    const items: (Terminal | Next)[] = []
    for (const { intent, target } of destinations) {
      if (intent.kind === 'end') {
        items.push(this.endTerminal(intent.hasValue, intent.value, activation.activationId))
      } else if (target !== undefined) {
        items.push({ type: 'next', elementId: target, input: intent.value })
      } else {
        items.push(this.exitTerminal(intent.action, intent.value, activation.activationId))
      }
    }
    return { failed: false, items }
  }

  private routeChild(
    scope: CompiledScope,
    scopeId: number,
    placement: CompiledPlacement,
    activation: Activation,
    terminals: readonly Terminal[],
  ): Success | FailureSignal {
    const items: (Terminal | Next)[] = []
    for (const terminal of terminals) {
      if (terminal.type === 'end') {
        items.push(terminal)
        continue
      }
      const routed = this.routeIntents(scope, scopeId, placement, activation, [
        { kind: 'emit', action: terminal.action, value: terminal.output, hasValue: true },
      ])
      if (routed.failed) return routed
      items.push(...routed.items)
    }
    return { failed: false, items }
  }

  private boundaryTerminals(intents: readonly Intent[], sourceActivationId: number): readonly Terminal[] {
    return Object.freeze(
      intents.map((intent) =>
        intent.kind === 'end'
          ? this.endTerminal(intent.hasValue, intent.value, sourceActivationId)
          : this.exitTerminal(intent.action, intent.value, sourceActivationId),
      ),
    )
  }

  private invalidRootExit(scope: CompiledScope, terminals: readonly Terminal[], runtimeScopeId: number): FailureSignal | undefined {
    if (scope.parentScopeId !== null) return undefined
    for (const terminal of terminals) {
      if (terminal.type === 'exit' && terminal.action !== null && !scope.exits.includes(terminal.action)) {
        return this.signal(
          this.failure(
            'unknown_action',
            `unknown root exit ${JSON.stringify(terminal.action)}`,
            runtimeScopeId,
            terminal.sourceActivationId,
            scope.ownerElementId,
          ),
          terminal.sourceActivationId,
          terminal.output,
        )
      }
    }
    return undefined
  }

  private callbackFailure(
    cause: unknown,
    ordinaryKind: FailureKind,
    failureMessage: string,
    scopeId: number,
    activationId: number,
    elementId: number,
    attempt: number | null,
    input: unknown,
    result: ScopeResult | null,
  ): FailureSignal {
    const kind = cause instanceof InvalidOutcome ? 'invalid_outcome' : ordinaryKind
    return this.signal(
      this.failure(
        kind,
        cause instanceof InvalidOutcome ? message(cause) : failureMessage,
        scopeId,
        activationId,
        elementId,
        attempt,
        cause,
      ),
      activationId,
      input,
      result,
    )
  }

  private scopeResult(terminals: readonly Terminal[]): ScopeResult {
    return Object.freeze({
      terminals,
      outputs: Object.freeze(terminals.filter((terminal) => terminal.hasOutput).map((terminal) => terminal.output)),
    })
  }

  private failure(
    kind: FailureKind,
    failureMessage: string,
    scopeId: number,
    activationId: number | null,
    elementId: number | null,
    attempt: number | null = null,
    cause: unknown = null,
    previous: Failure | null = null,
  ): Failure {
    return Object.freeze({
      failureId: ++this.nextFailureId,
      kind,
      message: failureMessage,
      cause,
      scopeId,
      activationId,
      elementId,
      attempt,
      previous,
    })
  }

  private signal(
    failure: Failure,
    activationId: number | null,
    input: unknown,
    result: ScopeResult | null = null,
    terminals: readonly Terminal[] = [],
  ): FailureSignal {
    return { failed: true, failure, activationId, input, result, terminals }
  }

  private activation(elementId: number, input: unknown): Activation {
    return { activationId: ++this.nextActivationId, elementId, input }
  }

  private endTerminal(hasOutput: boolean, output: unknown, sourceActivationId: number): EndTerminal {
    const shared = { type: 'end' as const, sequence: ++this.nextTerminalSequence, sourceActivationId }
    return hasOutput
      ? Object.freeze({ ...shared, hasOutput: true as const, output })
      : Object.freeze({ ...shared, hasOutput: false as const, output: undefined })
  }

  private exitTerminal(action: Action | null, output: unknown, sourceActivationId: number): ExitTerminal {
    return Object.freeze({
      type: 'exit',
      action,
      hasOutput: true,
      output,
      sequence: ++this.nextTerminalSequence,
      sourceActivationId,
    })
  }

  private scope(scopeId: number): CompiledScope {
    const scope = this.snapshot.scopes[scopeId - 1]
    if (scope === undefined) throw new Error('compiled scope is missing')
    return scope
  }

  private placement(elementId: number): CompiledPlacement {
    const placement = this.snapshot.placements[elementId - 1]
    if (placement === undefined) throw new Error('compiled placement is missing')
    return placement
  }
}

interface Activation {
  readonly activationId: number
  readonly elementId: number
  readonly input: unknown
}

interface Next {
  readonly type: 'next'
  readonly elementId: number
  readonly input: unknown
}

interface Success {
  readonly failed: false
  readonly items: readonly (Terminal | Next)[]
}

interface IntentSuccess {
  readonly failed: false
  readonly intents: readonly Intent[]
}

interface FailureSignal {
  readonly failed: true
  readonly failure: Failure
  readonly activationId: number | null
  readonly input: unknown
  readonly result: ScopeResult | null
  readonly terminals: readonly Terminal[]
}

interface ScopeSuccess {
  readonly failed: false
  readonly terminals: readonly Terminal[]
}

interface ScopeFailed {
  readonly failed: true
  readonly terminals: readonly Terminal[]
  readonly failure: Failure
}

type CallbackResult = { readonly ok: true; readonly intents: readonly Intent[] } | { readonly ok: false; readonly cause: unknown }

class Handle<State extends object> implements RunHandle<State> {
  readonly #promise: Promise<RunResult<State>>
  #done = false

  constructor(run: () => Promise<RunResult<State>>) {
    this.#promise = Promise.resolve()
      .then(run)
      .finally(() => {
        this.#done = true
      })
  }

  done(): boolean {
    return this.#done
  }

  result(): Promise<RunResult<State>> {
    return this.#promise
  }
}

function message(cause: unknown): string {
  return cause instanceof Error ? cause.message : String(cause)
}

File: python/caskada/_runner.py
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
# Copyright (c) 2026, Victor Duarte
from __future__ import annotations

import asyncio
import inspect
from dataclasses import dataclass
from typing import Any, TypeVar

from ._context import _Context, _Intent, _InvalidOutcome
from ._contracts import (
    Action,
    Completed,
    EndTerminal,
    ExitTerminal,
    Failed,
    Failure,
    FailureKind,
    RunResult,
    ScopeFailure,
    ScopeResult,
    Terminal,
)
from ._graph import (
    _CompiledPlacement,
    _CompiledScope,
    _CompiledSnapshot,
)

StateT = TypeVar("StateT")
_MISSING = object()


def _start(snapshot: _CompiledSnapshot, state: StateT) -> _Handle:
    loop = asyncio.get_running_loop()
    task = loop.create_task(_Run(snapshot, state).execute())
    return _Handle(task)


class _Run:
    def __init__(self, snapshot: _CompiledSnapshot, state: object) -> None:
        self.snapshot = snapshot
        self.state = state
        self.next_activation_id = 1  # The root Flow owns activation 1.
        self.next_failure_id = 0
        self.next_scope_id = 0
        self.next_terminal_sequence = 0

    async def execute(self) -> RunResult[Any]:
        outcome = await self._run_scope(1, 1, None)
        if isinstance(outcome, _ScopeFailed):
            return Failed(self.state, outcome.terminals, outcome.failure)
        return Completed(self.state, outcome.terminals)

    async def _run_scope(
        self,
        compiled_scope_id: int,
        owner_activation_id: int,
        input: object,
    ) -> _ScopeSuccess | _ScopeFailed:
        scope = self.snapshot.scope(compiled_scope_id)
        runtime_scope_id = self._new_scope_id()
        queue = [self._activation(scope.entry_element_id, input)]
        terminals: list[Terminal] = []
        started = 0
        failure: _FailureSignal | None = None

        while queue and failure is None:
            if scope.max_activations is not None and started >= scope.max_activations:
                blocked = queue[0]
                failure = _FailureSignal(
                    self._failure(
                        "activation_limit",
                        f"Flow {scope.name!r} exceeded max_activations",
                        runtime_scope_id,
                        blocked.activation_id,
                        blocked.element_id,
                    ),
                    blocked.activation_id,
                    blocked.input,
                )
                break

            width = min(scope.concurrency, len(queue))
            if scope.max_activations is not None:
                width = min(width, scope.max_activations - started)
            batch, queue = queue[:width], queue[width:]
            started += len(batch)
            outcomes = await asyncio.gather(
                *(self._run_activation(scope, runtime_scope_id, item) for item in batch)
            )
            for activation, outcome in zip(batch, outcomes, strict=True):
                if isinstance(outcome, _FailureSignal):
                    if failure is None:
                        failure = outcome
                    continue
                for item in outcome.items:
                    if isinstance(item, (EndTerminal, ExitTerminal)):
                        terminals.append(item)
                    elif failure is None:
                        queue.append(self._activation(item.element_id, item.input))

        ordered = tuple(sorted(terminals, key=lambda terminal: terminal.sequence))
        if failure is not None:
            return await self._recover_scope(
                scope,
                runtime_scope_id,
                owner_activation_id,
                failure,
                ordered,
            )

        if scope.combine is None:
            return _ScopeSuccess(ordered)
        result = ScopeResult(ordered)
        callback = await self._callback(scope.combine, input, result)
        if isinstance(callback, Exception):
            signal = self._callback_failure(
                callback,
                "flow_combine",
                "Flow combine failed",
                runtime_scope_id,
                owner_activation_id,
                scope.owner_element_id,
                None,
                input,
                result=result,
            )
            return await self._recover_scope(
                scope,
                runtime_scope_id,
                owner_activation_id,
                signal,
                ordered,
            )
        if not callback:
            return _ScopeSuccess(ordered)

        transformed = self._boundary_terminals(callback, owner_activation_id)
        invalid = self._invalid_root_exit(scope, transformed, runtime_scope_id)
        if invalid is not None:
            return await self._recover_scope(
                scope,
                runtime_scope_id,
                owner_activation_id,
                invalid,
                ordered,
            )
        return _ScopeSuccess(transformed)

    async def _run_activation(
        self,
        scope: _CompiledScope,
        runtime_scope_id: int,
        activation: _Activation,
    ) -> _Success | _FailureSignal:
        placement = self.snapshot.placement(activation.element_id)
        if placement.kind == "flow":
            if placement.owned_scope_id is None:
                raise RuntimeError("compiled Flow has no owned scope")
            # ponytail: recurse by Flow depth; use an explicit stack only if real
            # workflows reach Python's recursion limit.
            child = await self._run_scope(
                placement.owned_scope_id,
                activation.activation_id,
                activation.input,
            )
            if isinstance(child, _ScopeFailed):
                return _FailureSignal(
                    child.failure, activation.activation_id, activation.input
                )
            return self._route_child(
                scope, runtime_scope_id, placement, activation, child.terminals
            )

        intents = await self._run_node(placement, runtime_scope_id, activation)
        if isinstance(intents, _FailureSignal):
            return intents
        return self._route_intents(
            scope, runtime_scope_id, placement, activation, intents
        )

    async def _run_node(
        self,
        placement: _CompiledPlacement,
        scope_id: int,
        activation: _Activation,
    ) -> tuple[_Intent, ...] | _FailureSignal:
        if placement.handler is None or placement.retry is None:
            raise RuntimeError("compiled Node is incomplete")
        previous: Failure | None = None

        for attempt in range(1, placement.retry.max_attempts + 1):
            callback = await self._callback(placement.handler, activation.input)
            if not isinstance(callback, Exception):
                return callback or (_Intent("emit", None, activation.input, True),)

            kind: FailureKind = (
                "invalid_outcome"
                if isinstance(callback, _InvalidOutcome)
                else "handler"
            )
            failure = self._failure(
                kind,
                str(callback),
                scope_id,
                activation.activation_id,
                placement.element_id,
                attempt,
                callback,
                previous,
            )
            if kind == "handler" and attempt < placement.retry.max_attempts:
                policy_failure = self._retry(placement, attempt, failure)
                if isinstance(policy_failure, Failure):
                    return _FailureSignal(
                        policy_failure, activation.activation_id, activation.input
                    )
                if policy_failure is not None:
                    if policy_failure:
                        await asyncio.sleep(policy_failure / 1000)
                    previous = failure
                    continue
            return await self._recover_node(placement, scope_id, activation, failure)

        raise RuntimeError("retry loop did not return")

    def _retry(
        self,
        placement: _CompiledPlacement,
        attempt: int,
        failure: Failure,
    ) -> int | None | Failure:
        if placement.retry is None:
            raise RuntimeError("compiled Node has no retry policy")
        try:
            should_retry = placement.retry.should_retry(failure)
            if type(should_retry) is not bool:
                raise TypeError("should_retry must return bool")
            if not should_retry:
                return None
            delay = placement.retry.delay_ms
            delay_ms = delay(attempt, failure) if callable(delay) else delay
            if (
                not isinstance(delay_ms, int)
                or isinstance(delay_ms, bool)
                or delay_ms < 0
            ):
                raise TypeError("delay_ms must return a nonnegative integer")
            return delay_ms
        except Exception as cause:  # noqa: BLE001 - retry policy is user code
            return self._failure(
                "retry_policy",
                str(cause),
                failure.scope_id,
                failure.activation_id,
                failure.element_id,
                failure.attempt,
                cause,
                failure,
            )

    async def _recover_node(
        self,
        placement: _CompiledPlacement,
        scope_id: int,
        activation: _Activation,
        failure: Failure,
    ) -> tuple[_Intent, ...] | _FailureSignal:
        if placement.recover is None:
            return _FailureSignal(failure, activation.activation_id, activation.input)
        callback = await self._callback(placement.recover, activation.input, failure)
        if isinstance(callback, Exception):
            kind: FailureKind = (
                "invalid_outcome"
                if isinstance(callback, _InvalidOutcome)
                else "node_recovery"
            )
            replacement = self._failure(
                kind,
                str(callback),
                scope_id,
                activation.activation_id,
                placement.element_id,
                None,
                callback,
                failure,
            )
            return _FailureSignal(
                replacement, activation.activation_id, activation.input
            )
        if not callback:
            return _FailureSignal(failure, activation.activation_id, activation.input)
        return callback

    async def _recover_scope(
        self,
        scope: _CompiledScope,
        runtime_scope_id: int,
        owner_activation_id: int,
        signal: _FailureSignal,
        terminals: tuple[Terminal, ...],
    ) -> _ScopeSuccess | _ScopeFailed:
        if scope.recover is None:
            return _ScopeFailed(terminals, signal.failure)
        scope_failure = ScopeFailure(
            signal.failure,
            terminals,
            signal.result,
            signal.activation_id,
        )
        callback = await self._callback(scope.recover, signal.input, scope_failure)
        if isinstance(callback, Exception):
            kind: FailureKind = (
                "invalid_outcome"
                if isinstance(callback, _InvalidOutcome)
                else "flow_recovery"
            )
            return _ScopeFailed(
                terminals,
                self._failure(
                    kind,
                    str(callback),
                    runtime_scope_id,
                    owner_activation_id,
                    scope.owner_element_id,
                    None,
                    callback,
                    signal.failure,
                ),
            )
        if not callback:
            return _ScopeFailed(terminals, signal.failure)

        transformed = self._boundary_terminals(callback, owner_activation_id)
        invalid = self._invalid_root_exit(scope, transformed, runtime_scope_id)
        if invalid is not None:
            return _ScopeFailed(terminals, invalid.failure)
        return _ScopeSuccess(transformed)

    async def _callback(
        self,
        callback: Any,
        input: object,
        extra: object = _MISSING,
    ) -> tuple[_Intent, ...] | Exception:
        context = _Context(self.state, input)
        try:
            value = callback(context) if extra is _MISSING else callback(context, extra)
            if inspect.isawaitable(value):
                value = await value
            if value is not None:
                raise _InvalidOutcome("callbacks must return None")
            return context.close()
        except Exception as cause:  # noqa: BLE001 - callback is user code
            context.close()
            return cause

    def _route_intents(
        self,
        scope: _CompiledScope,
        scope_id: int,
        placement: _CompiledPlacement,
        activation: _Activation,
        intents: tuple[_Intent, ...],
    ) -> _Success | _FailureSignal:
        destinations: list[tuple[_Intent, int | None]] = []
        for intent in intents:
            if intent.kind == "end":
                destinations.append((intent, None))
                continue
            target = next(
                (
                    link.target_element_id
                    for link in placement.links
                    if link.action == intent.action
                ),
                None,
            )
            if (
                target is None
                and intent.action is not None
                and intent.action not in scope.exits
            ):
                return _FailureSignal(
                    self._failure(
                        "unknown_action",
                        f"unknown action {intent.action!r} from {placement.name!r}",
                        scope_id,
                        activation.activation_id,
                        placement.element_id,
                    ),
                    activation.activation_id,
                    activation.input,
                )
            destinations.append((intent, target))

        items: list[Terminal | _Next] = []
        for intent, target in destinations:
            if intent.kind == "end":
                items.append(
                    self._end_terminal(
                        intent.has_value, intent.value, activation.activation_id
                    )
                )
            elif target is not None:
                items.append(_Next(target, intent.value))
            else:
                items.append(
                    self._exit_terminal(
                        intent.action, intent.value, activation.activation_id
                    )
                )
        return _Success(tuple(items))

    def _route_child(
        self,
        scope: _CompiledScope,
        scope_id: int,
        placement: _CompiledPlacement,
        activation: _Activation,
        terminals: tuple[Terminal, ...],
    ) -> _Success | _FailureSignal:
        items: list[Terminal | _Next] = []
        for terminal in terminals:
            if isinstance(terminal, EndTerminal):
                items.append(terminal)
                continue
            intent = _Intent("emit", terminal.action, terminal.output, True)
            routed = self._route_intents(
                scope, scope_id, placement, activation, (intent,)
            )
            if isinstance(routed, _FailureSignal):
                return routed
            items.extend(routed.items)
        return _Success(tuple(items))

    def _boundary_terminals(
        self,
        intents: tuple[_Intent, ...],
        source_activation_id: int,
    ) -> tuple[Terminal, ...]:
        terminals: list[Terminal] = []
        for intent in intents:
            if intent.kind == "end":
                terminals.append(
                    self._end_terminal(
                        intent.has_value, intent.value, source_activation_id
                    )
                )
            else:
                terminals.append(
                    self._exit_terminal(
                        intent.action, intent.value, source_activation_id
                    )
                )
        return tuple(terminals)

    def _invalid_root_exit(
        self,
        scope: _CompiledScope,
        terminals: tuple[Terminal, ...],
        runtime_scope_id: int,
    ) -> _FailureSignal | None:
        if scope.parent_scope_id is not None:
            return None
        for terminal in terminals:
            if (
                isinstance(terminal, ExitTerminal)
                and terminal.action is not None
                and terminal.action not in scope.exits
            ):
                return _FailureSignal(
                    self._failure(
                        "unknown_action",
                        f"unknown root exit {terminal.action!r}",
                        runtime_scope_id,
                        terminal.source_activation_id,
                        scope.owner_element_id,
                    ),
                    terminal.source_activation_id,
                    terminal.output,
                )
        return None

    def _callback_failure(
        self,
        cause: Exception,
        ordinary_kind: FailureKind,
        message: str,
        scope_id: int,
        activation_id: int,
        element_id: int,
        attempt: int | None,
        input: object,
        *,
        result: ScopeResult | None = None,
    ) -> _FailureSignal:
        kind: FailureKind = (
            "invalid_outcome" if isinstance(cause, _InvalidOutcome) else ordinary_kind
        )
        return _FailureSignal(
            self._failure(
                kind,
                str(cause) if isinstance(cause, _InvalidOutcome) else message,
                scope_id,
                activation_id,
                element_id,
                attempt,
                cause,
            ),
            activation_id,
            input,
            result,
        )

    def _failure(
        self,
        kind: FailureKind,
        message: str,
        scope_id: int,
        activation_id: int | None,
        element_id: int | None,
        attempt: int | None = None,
        cause: BaseException | None = None,
        previous: Failure | None = None,
    ) -> Failure:
        self.next_failure_id += 1
        return Failure(
            self.next_failure_id,
            kind,
            message,
            cause,
            scope_id,
            activation_id,
            element_id,
            attempt,
            previous,
        )

    def _activation(self, element_id: int, input: object) -> _Activation:
        self.next_activation_id += 1
        return _Activation(self.next_activation_id, element_id, input)

    def _new_scope_id(self) -> int:
        self.next_scope_id += 1
        return self.next_scope_id

    def _end_terminal(
        self, has_output: bool, output: object, source_activation_id: int
    ) -> EndTerminal:
        self.next_terminal_sequence += 1
        return EndTerminal(
            has_output,
            output,
            self.next_terminal_sequence,
            source_activation_id,
        )

    def _exit_terminal(
        self, action: Action | None, output: object, source_activation_id: int
    ) -> ExitTerminal:
        self.next_terminal_sequence += 1
        return ExitTerminal(
            action,
            output,
            self.next_terminal_sequence,
            source_activation_id,
        )


@dataclass(frozen=True, slots=True)
class _Activation:
    activation_id: int
    element_id: int
    input: object


@dataclass(frozen=True, slots=True)
class _Next:
    element_id: int
    input: object


@dataclass(frozen=True, slots=True)
class _Success:
    items: tuple[Terminal | _Next, ...]


@dataclass(frozen=True, slots=True)
class _FailureSignal:
    failure: Failure
    activation_id: int | None
    input: object
    result: ScopeResult | None = None


@dataclass(frozen=True, slots=True)
class _ScopeSuccess:
    terminals: tuple[Terminal, ...]


@dataclass(frozen=True, slots=True)
class _ScopeFailed:
    terminals: tuple[Terminal, ...]
    failure: Failure


class _Handle:
    def __init__(self, task: asyncio.Task[RunResult[Any]]) -> None:
        self._task = task

    def done(self) -> bool:
        return self._task.done()

    async def result(self) -> RunResult[Any]:
        return await self._task
